### ISSUE TITLE
Add a way for users to specify custom facet builder method for all integrations 

### ISSUE THREAD

**Issue by mobuchowski** [OpenLineage/OpenLineage#3144]
@anandhimurali added feature to Airflow OpenLineage integration that allows users to specify methods that produce custom facets
https://github.com/apache/airflow/pull/38982

Similarly, we have [CustomFacetBuilder](https://github.com/OpenLineage/OpenLineage/blob/31f8ce588526e9c7c4bc7d849699cb7ce2969c8f/integration/spark/shared/src/main/java/io/openlineage/spark/api/CustomFacetBuilder.java#L34) in Spark. 

However, those are integration specific methods. It does not work for other integrations, like Flink or dbt, and the mechanisms are custom to the integration. It would be nice to have a way to specify a method, configure OpenLineage client to point at it, and have it produce custom metadata, regardless being in different integrations. 

On the other hand, lack of any integration-specific argument means users aren't able to refer to any run, job, dataset specific data, in contrast to integration specific ones. For example, Spark integration's CustomFacetBuilder allows you to get particular LogicalPlan for given QueryExecution and extract data from it.
Maybe the better option would be to add specific way of integration to all OL-maintained integrations (Flink, dbt...) and think of it as one of the core features we're trying to provide, rather than have something totally generic?

Another option would be to define interface differently. The "producer" of custom facet could be a class, and it could have Spark-inspired `isDefinedAt` method, which we'd call the with potential arguments, and based on the response we'd figure out whether it supports the given arguments. This would work for any sufficiently generic language - Python, Java, Go - but would require users to use reflection to get the particular data if needed.

**Comment by pawel-big-lebowski** [comment]
I see two use cases for custom facets that are NOT integration specific:
 * add facets based on static values provided in OL config,
 * add facets based on environment variables.
 
Are there any other potential use cases?

**Comment by MassyB** [comment]
@pawel-big-lebowski  +1 on the facets from env vars. We need that for DBT for example. 
Giving the dbt perspective here. A useful arg to the callback could be the path the dbt artifacts.

**Comment by mobuchowski** [comment]
@MassyB yes - that's the intention. Not only for dbt, but also any custom integration based on OpenLineage clients.

**Comment by mobuchowski** [comment]
@harels @julienledem @pawel-big-lebowski @JDarDagran 

Here, we are thinking about "adding" facets. 
What do you think about having generic "event mutation hook"? The mechanism could be called after integration prepared the event, and before it's send to transports.

The argument would be the event and we'd expect to get valid event out of the function.

**Comment by JDarDagran** [comment]
> What do you think about having generic "event mutation hook"? The mechanism could be called after integration prepared the event, and before it's send to transports.

In general, I like the idea. Should the event be validated somehow before final emit step? Or is it user's _fault_ when the event is not OL-compliant after going through mutation hooks.
Would you like to share broader picture how you'd see those hooks? E.g.:
* How to deliver such hooks in Java? Through ServiceLoader?
* Would only one hook be allowed per client setup? If not, should they be ordered?
* Should failed mutation cause failure of the whole emission or just skip?

Probably there's more when it comes to implementation but I'm curious at the moment to follow if this direction is right at all.

**Comment by pawel-big-lebowski** [comment]
@mobuchowski I like the hook idea. It gives great agility for the producers. Additionally, I would allow discarding the event through the hook.

**Comment by harels** [comment]
I think this is a good idea.
I agree with @JDarDagran 's concern and I would opt to add a default behavior that validates this is still a valid OL event after applying the mutation, and error out if it's not.
I would also add a "permissive" flag (non default) to override this behavior.
